spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindPotentialSecurityCheckBasedOnUntrustedSource.java

```diff
@@ -0,0 +1,242 @@
+    private Map<XMethod, Set<CallerInfo>> methodsCalledInsidePrivilegedAction = new HashMap<>();
+
+    private String topParameterName = null;
+
+    private boolean isDoPrivileged = false;
+    private boolean isDoPrivilegedRun = false;
+
+    BugAccumulator bugAccumulator;
+
+    public FindPotentialSecurityCheckBasedOnUntrustedSource(BugReporter bugReporter) {
+        this.bugAccumulator = new BugAccumulator(bugReporter);
+    }
+
+    @Override
+    public void visit(JavaClass obj) {
+        isDoPrivileged = Subtypes2.instanceOf(getDottedClassName(), "java.security.PrivilegedAction")
+                || Subtypes2.instanceOf(getDottedClassName(), "java.security.PrivilegedExceptionAction");
+    }
+
+    @Override
+    public void visit(Method obj) {
+        nonFinalMethodsCalledOnParam.clear();
+        isDoPrivilegedRun = isDoPrivileged && "run".equals(getMethodName());
+    }
+
+    @Override
+    public void visit(Code obj) {
+        if (!isDoPrivilegedRun && (!getThisClass().isPublic() || !getMethod().isPublic())) {
+            return;
+        }
+        super.visit(obj);
+    }
+
+    @Override
+    public void visitAfter(JavaClass obj) {
+        bugAccumulator.reportAccumulatedBugs();
+    }
+
+    @Override
+    public void sawOpcode(int seen) {
+        if (seen == Const.INVOKEVIRTUAL && getXClassOperand() != null && getXMethodOperand() != null) {
+            if (getXMethodOperand().getSignature().endsWith("V")) {
+                return;
+            }
+            OpcodeStack.Item object = stack.getStackItem(0);
+
+            if (isDoPrivilegedRun) {
+                if (!getXClassOperand().isFinal() && !getXMethodOperand().isFinal() && object.isInitialParameter()
+                        && isNestingMethodLocalVariable(object)) {
+                    addToMethodsCalledInsidePrivilegedAction(getXMethodOperand(), object);
+                }
+                return;
+            }
+
+            if (getXClass().isPublic() && getXMethod().isPublic() && !getXClassOperand().isFinal()
+                    && !getXMethodOperand().isFinal() && object.isInitialParameter() && object.getXField() == null
+                    && topParameterName != null) {
+                addTononFinalMethodsCalledOnParam(getClassDescriptorOperand(), getXMethodOperand(), object);
+            }
+        } else if (seen == Const.INVOKESTATIC && getXMethodOperand() != null
+                && "doPrivileged".equals(getXMethodOperand().getName())) {
+            OpcodeStack.Item action = stack.getStackItem(0);
+            CallPair callPair = lookForCalledOutsideAndInside(action);
+            if (callPair != null) {
+                reportBug(callPair);
+            }
+        }
+    }
+
+    private boolean isNestingMethodLocalVariable(OpcodeStack.Item object) {
+        XField field = object.getXField();
+        if (field == null) {
+            return false;
+        }
+
+        return field.getName().matches("val\\$.*");
```

Comment:
This [`String#matches(String)` method](https://docs.oracle.com/javase/8/docs/api/java/lang/String.html#matches-java.lang.String-) [compiles Regex `Pattern` every time](https://github.com/openjdk/jdk/blob/jdk8-b120/jdk/src/share/classes/java/util/regex/Pattern.java#L1133).

For better performance, consider compile a `Pattern` instance in advance, keep it in an instance field, and reuse it in this method.